### PR TITLE
Extend query depth limiter allowing for more detailed rules

### PR DESCRIPTION
**Pull request by tsmith023**
Enhances the query depth limiter by providing options to specify additional rules for fields within the GraphQL query.
These can be the field name on its own or including the field arguments.

## Description

Exposes the new `should_ignore` argument to `QueryDepthLimiter` that takes a single argument of type `IgnoreContext` allowing the user greater verbosity when defining their rules. This dataclass can be extended to provide more ignore options by future PRs in response to further feature requests driven by specific use cases.

This functionality does not replace previous functionality but simply extends it. So you can still use the `QueryDepthLimiter` extension as it was previously implemented but now you can also use the further verbose functionality. However, the previous implementation is deprecated and will be removed in future releases.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry/issues/2338

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Tommy Smith for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by tsmith023** [comment]
Linking the previous review here: https://github.com/strawberry-graphql/strawberry/pull/2501#discussion_r1093074149 :grin:

**Comment by tsmith023** [comment]
@patrick91, this PR is in a state to be reviewed :D

File: docs/extensions/query-depth-limiter.md [hunk 1]
```diff
@@ -164,3 +223,90 @@ schema.execute(
 ```
 
 </details>
+
+<details>
+  <summary>Ignoring fields with FieldRule and arguments</summary>
+
+<!-- Cannot specify python as code for this block as blacken-docs fails
+to parse the field_arguments line within FieldRule -->
```
Comment (patrick91) [inline docs/extensions/query-depth-limiter.md:231]: this is odd, I'll check it soon 😊

**Review by patrick91 — COMMENTED** [review]
Can you tell me more about this feature? what's your use case?

Also in terms of API, I wonder if we could use callables instead of field rules. I think that might make our implementation shorter, and potentially the api could be easier to use 🤔

**Comment by tsmith023** [comment]
The use case is that described in https://github.com/strawberry-graphql/strawberry/issues/2338, which allows to limit the depth of queries depending on the arguments of specific fields in addition to the names of said fields.

Let's have a discussion in Discord on the implementation detail of callables instead of FieldRules as I'm not 100% clear on how that would look in the code :grin:

File: docs/extensions/query-depth-limiter.md (hunk 1 above)
Comment (tsmith023) [reply docs/extensions/query-depth-limiter.md:231]: I've changed that implementation so there's no need to investigate this anymore! :grin:

**Comment by tsmith023** [comment]
@patrick91, this PR has been refactored to take into account the callables discussion we had previously. If you have time to review it then it should be ready to merge!

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+Release type: patch
```
Comment (patrick91) [inline RELEASE.md:1]:
```suggestion
Release type: minor
```

I think this is a new feature 😊

File: RELEASE.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+This patch introduces the new `should_ignore` argument to the `QueryDepthLimiter` extension that provides
```
Comment (patrick91) [inline RELEASE.md:3]:
```suggestion
This release introduces the new `should_ignore` argument to the `QueryDepthLimiter` extension that provides
```

File: strawberry/extensions/query_depth_limiter.py [hunk 4]
```diff
@@ -71,28 +101,47 @@ class QueryDepthLimiter(AddValidationRules):
...
-        validator = create_validator(max_depth, ignore, callback)
+        if should_ignore is not None:
+            if not callable(should_ignore):
+                raise TypeError(
+                    "The `should_ignore` argument to "
+                    "`QueryDepthLimiter` must be a callable."
+                )
+            validator = create_validator(max_depth, should_ignore, callback)
+        else:
+            warnings.warn(
```
Comment (patrick91) [inline strawberry/extensions/query_depth_limiter.py:132]: thanks for adding the deprecation 😊

File: strawberry/extensions/query_depth_limiter.py [hunk 5]
```diff
@@ -28,27 +28,57 @@
...
 
-if TYPE_CHECKING:
-    from graphql.language import DefinitionNode, Node
-    from graphql.validation import ValidationContext
+IgnoreType = Union[Callable[[str], bool], re.Pattern, str]
+
+FieldArgumentType = Union[
+    bool, int, float, str, List["FieldArgumentType"], Dict[str, "FieldArgumentType"]
+]
+FieldArgumentsType = Dict[str, FieldArgumentType]
```
Comment (patrick91) [inline strawberry/extensions/query_depth_limiter.py:70]: we might want to move these out at some point, or reuse something classes we already have

**Review by patrick91 — APPROVED** [review]

**Pull request merged**